### PR TITLE
chorse: change numpy to jax numpy

### PR DESCRIPTION
**Pull request by pphuc25**
Hi,
I just do a simple remove numpy and change to jax numpy, seen at now jax is more robust and can replace to numpy.
I would like to cc @stevhliu to review my PR.

### CODE REVIEW

**Comment by stevhliu** [comment]
Makes sense to me to use the built-in NumPy module, but I'm not too familiar with how robust JAX is compared to NumPy. What do you think @LysandreJik ?

**Comment by sanchit-gandhi** [comment]
This is not such a great idea! Let me explain why: any operations run as `numpy` code will be executed on the CPU. Any operation run as `jax.numpy` code will be executed on the **accelerator device** (GPU or TPU). Note that this behaviour in JAX is different to PyTorch: by default, JAX code will be run on the accelerator device (if available), whereas in PyTorch it will be the CPU (unless specified otherwise).

This is significant because JAX uses [*asynchronous dispatch*](https://jax.readthedocs.io/en/latest/async_dispatch.html) to coordinate code between the CPU and accelerator device. If the accelerator device is running `jax.numpy` code, the CPU is free to carry on and run subsequent `numpy` code. It is not blocked by the accelerator running `jax` code, and can in fact already execute any 'future' code. This is useful in ML applications, since the accelerator can stay busy running the modelling code (forward and backward pass), whilst the CPU can go ahead and prepare the next batch of data ahead of time.

If we change our data preparation to use `jax.numpy` code, then we execute it on the accelerator device. This means we have to wait until the accelerator is free before running the data preparation operations, so we loose our asynchronous dispatch advantage.

An easy rule to keep in mind when writing JAX code is that **only the modelling code should be in JAX**. Everything else should be in NumPy. This way, we ensure only the modelling code is executed on the accelerator device, and everything else on CPU, thus maximising the async dispatch potential.

Let me know if you have any questions @pphuc25 - more than happy to answer!

**Comment by pphuc25** [comment]
Hmm, that's sound great insight, this is the first time I hear about it, very helpful information for me, thank you so much
